lua/typescript-tools/api.lua

```diff
@@ -149,6 +149,46 @@ function M.request_diagnostics(callback)
   }, callback)
 end
 
+---@param is_sync boolean
+function M.rename_file(is_sync)
+  local typescript_client = get_typescript_client(0)
+  if typescript_client == nil then
+    return
+  end
+
+  local source = vim.api.nvim_buf_get_name(0)
+
+  local params = {
+    files = {
+      {
+        oldUri = "file://" .. source,
+      },
+    },
+  }
+
+  vim.ui.input({ prompt = "New path: ", default = source }, function(input)
+    if input == "" or input == source or input == nil then
+      return
+    end
+
+    params.files[1].newUri = "file://" .. input
+
+    if is_sync then
+      local res = typescript_client.request_sync(c.LspMethods.WillRenameFiles, params, timeout, 0)
+      if not res.err then
+        vim.lsp.util.apply_workspace_edit(res.result, "utf-8")
+      end
+    else
+      typescript_client.request(c.LspMethods.WillRenameFiles, params, function(err, res)
+        if not err then
+          vim.lsp.util.apply_workspace_edit(res, "utf-8")
+        end
+      end, 0)
+      vim.api.nvim_buf_set_name(0, input)
```

Comment:
Given this scenario:
1. Open the file `test.ts`
2. Run `TSToolsRenameFile` with input `test1.ts`
3. run `:w`
We have two files on the file disc `test.ts` and `test1.ts`. I'm wondering if we could use `vim.loop.fs_rename(source, target)` and actually rename the file on the file disc?

Reply:
Yes, absolutely. I've just added this line and it worked :+1: Yesterday I thought that `nvim_buf_set_name` will rename the file in the background, but that only renamed filename for the buffer, not the file itself. :facepalm: